File: src/main/java/com/dersaun/apigestaocontas/infra/repository/FormaPagamentoRepository.java
package com.dersaun.apigestaocontas.infra.repository;

import com.dersaun.apigestaocontas.domain.FormaPagamento;
import com.dersaun.apigestaocontas.domain.Pessoa;
import org.eclipse.collections.api.list.MutableList;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.lang.Nullable;

public interface FormaPagamentoRepository extends JpaRepository<FormaPagamento, Long> {

    @EntityGraph(attributePaths = {"dono"})
    MutableList<FormaPagamento> findAllByAtivoIsTrueOrderByNomeAsc();

    @EntityGraph(attributePaths = {"dono"})
    MutableList<FormaPagamento> findAllByDonoIdAndAtivoIsTrueOrderByNomeAsc(Long idPessoa);

    /*and (
    ( :pessoa is not null and ( fp.dono = :pessoa or fp.id in (6L, 7L) ) ) or (:pessoa is null)
    )*/
    @Query("""
            from FormaPagamento fp 
            left join fetch fp.dono dono 
            where exists (
                select fi.id
                from FaturaItem fi
                where fi.lancamento.formaPagamento = fp
                and fi.fechamento = :mesReferencia
                and fi.ano = :anoReferencia
                and fi.lancamento.pessoa = COALESCE(:pessoa, fi.lancamento.pessoa) 
            )
            order by fp.nome ASC
                
            """)
    MutableList<FormaPagamento> buscarFormasPagamentosComCompras(@Param("mesReferencia") Long mesReferencia,
                                                                 @Param("anoReferencia") Integer anoReferencia,
                                                                 @Param("pessoa") @Nullable Pessoa pessoa);
}

File: src/app/conta/repositories/forma-pagamento.repository.ts
import { PessoaUsuarioAutenticadoDTO } from '@app/autenticacao/dtos/usuario-autenticado.dto';
import { FaturaItem } from '@app/fatura/models/fatura-item.entity';
import { Pessoa } from '@app/pessoa/models/pessoa.entity';
import { Injectable } from '@nestjs/common';
import { EntityManager, Repository } from 'typeorm';
import { IFormaPagamentoRepository } from '../interfaces/forma-pagamento.repository.interface';
import { FormaPagamento } from '../models/forma-pagamento.entity';

@Injectable()
export class FormaPagamentoRepository
  extends Repository<FormaPagamento>
  implements IFormaPagamentoRepository
{
  constructor(em: EntityManager) {
    super(FormaPagamento, em);
  }
  async buscarFormasPagamentosAtivasAdmin(): Promise<FormaPagamento[]> {
    return await this.find({ where: { ativo: true }, order: { nome: 'ASC' } });
  }
  async buscarFormasPagamentosDono(
    idPessoa: number,
  ): Promise<FormaPagamento[]> {
    return await this.find({
      where: {
        ativo: true,
        dono: {
          id: idPessoa,
        },
      },
      order: { nome: 'ASC' },
      relations: ['dono'],
    });
  }
  async buscarFormasPagamentosComCompras(
    mesReferencia: number,
    anoReferencia: number,
    pessoa?: Pessoa | PessoaUsuarioAutenticadoDTO,
  ): Promise<FormaPagamento[]> {
    const builder = this.createQueryBuilder('fp');

    const query = builder.leftJoinAndSelect('fp.dono', 'dono').whereExists(
      builder
        .subQuery()
        .from(FaturaItem, 'fi')
        .leftJoin('fi.lancamento', 'lancamento')
        .leftJoin('lancamento.formaPagamento', 'formaPagamento')
        .leftJoin('lancamento.pessoa', 'pessoa')
        .where('formaPagamento.id = fp.id')
        .andWhere('fi.fechamento = :mesReferencia', {
          mesReferencia,
        })
        .andWhere('fi.ano = :anoReferencia', {
          anoReferencia,
        })
        .andWhere('pessoa.id = coalesce(:idPessoa, pessoa.id)', {
          idPessoa: pessoa?.id,
        }),
    );
    const formasPagamentos = await query.orderBy('fp.nome', 'ASC').getMany();
    return formasPagamentos;
  }
}
